pysmt/solvers/optimizer.py

```diff
@@ -0,0 +1,243 @@
+
+
+
+class SUAOptimizerMixin(Optimizer):
+    """Optimizer mixin using solving under assumptions"""
+
+    def _lt(self, x, y):
+        otype = self.environment.stc.get_type(x)
+        mgr = self.environment.formula_manager
+        if otype.is_int_type() or otype.is_real_type():
+            return mgr.LT(x, y)
+        elif otype.is_bv_type():
+            return mgr.BVULT(x, y)
+
+    def _le(self, x, y):
+        otype = self.environment.stc.get_type(x)
+        mgr = self.environment.formula_manager
+        if otype.is_int_type() or otype.is_real_type():
+            return mgr.LE(x, y)
+        elif otype.is_bv_type():
+            return mgr.BVULE(x, y)
+
+    def optimize(self, cost_function, initial_cost=None, callback=None):
+        last_model = None
+        keep = True
+
+        initial_assumptions = []
+        if initial_cost is not None:
+            initial_assumptions.append(self._lt(cost_function, initial_cost))
+
+        cost_so_far = None
+        while keep:
+            if cost_so_far is not None:
+                k = self._lt(cost_function, cost_so_far)
+                keep = self.solve(assumptions=[k])
+            else:
+                keep = self.solve(assumptions=initial_assumptions)
+
+            if keep:
+                last_model = self.get_model()
+                cost_so_far = self.get_value(cost_function)
+                if callback is not None:
+                    exit_request = callback(last_model)
+                    if exit_request is True:
+                        break
+        return last_model
+
+    def pareto_optimize(self, cost_functions, callback=None):
+        mgr = self.environment.formula_manager
+
+        eliminated = []
+        terminated = False
+        while not terminated:
+            last_model = None
+            keep = True
+            costs_so_far = None
+            while keep:
+                if costs_so_far is not None:
+                    k = [self._le(cost_functions[i], costs_so_far[i])
+                         for i in range(len(cost_functions))]
+                    k.append(mgr.Or(self._lt(cost_functions[i], costs_so_far[i])
+                                    for i in range(len(cost_functions))))
+                    keep = self.solve(assumptions=eliminated + k)
+                else:
+                    keep = self.solve(assumptions=eliminated)
+
+                if keep:
+                    last_model = self.get_model()
+                    costs_so_far = []
+                    for i in range(len(cost_functions)):
+                        costs_so_far.append(self.get_value(cost_functions[i]))
+                    if callback is not None:
+                        exit_request = callback(last_model)
+                        if exit_request is True:
+                            break
+            if last_model is not None:
+                yield last_model
+                eliminated.append(mgr.Or(self._lt(cost_functions[i],
```

Comment:
Isn't this missing a Not?

Reply:
No, we want to find a new model in which at least one criterion is LT wrt the current value, that is sufficient to prune the model we just found